pkg/operator/metacomponentcontroller/component_controller.go

```diff
@@ -0,0 +1,117 @@
+package metacomponentcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/eparis/bugzilla"
+	"github.com/openshift-eng/shodan/pkg/operator/bugutil"
+	"github.com/openshift-eng/shodan/pkg/operator/config"
+	"github.com/openshift-eng/shodan/pkg/operator/controller"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// MetaComponentController purpose is to search for NEW bugs that has Developer Whiteboard field populated with "meta-component" names.
+// When such bug is found, and it is assigned to default component owner, this controller will reassign that bug to "meta-component" owner.
+// For example, if an SingleNode keyword is found in Developer Whiteboard and the bug is NEW and assigned to Stefan, this controller will reassign
+// this bug automatically to SNO engineer and move it to ASSIGNED.
+type MetaComponentController struct {
+	controller.ControllerContext
+	config config.OperatorConfig
+}
+
+func NewMetaComponentController(ctx controller.ControllerContext, operatorConfig config.OperatorConfig, recorder events.Recorder) factory.Controller {
+	c := &MetaComponentController{
+		ControllerContext: ctx,
+		config:            operatorConfig,
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(3*time.Hour).ToController("MetaComponentController", recorder)
+}
+
+func (c *MetaComponentController) sync(ctx context.Context, context factory.SyncContext) error {
+	client := c.NewBugzillaClient(ctx)
+	slackClient := c.SlackClient(ctx)
+
+	// if no config is found, fail fast
+	if c.config.MetaComponents == nil {
+		return nil
+	}
+
+	result, err := client.Search(getBugsQuery(c.config.Components.List(), c.config.MetaComponents.List()))
+	if err != nil {
+		return err
+	}
+
+	bugsToUpdate := map[int]bugzilla.BugUpdate{}
+
+	for i := range result {
+		for name, component := range c.config.MetaComponents {
+			// we only looking for bugs with meta component name inside developer whiteboard
+			// and bugs that does not have meta component lead already assigned to them.
+			// NOTE: the search result only search for NEW bugs, so ASSIGNED or any other bugs are not present here.
+			if !strings.Contains(result[i].DevelWhiteboard, name) || result[i].AssignedTo == component.Lead {
```

Comment:
so in order to intentionally assign to the (non-meta-)component lead you have to remove the keyword from the whiteboard, right? In other words: you are checking for the state of the bug, not an "assign to lead" event.

Reply:
or set to ASSIGNED.